Detect more `cfg`d out items in resolution errors

estebank opened the pull request:
Use a visitor to collect *all* items (including those nested) that were stripped behind a `cfg` condition.

```
error[E0425]: cannot find function `f` in this scope
  --> $DIR/nested-cfg-attrs.rs:4:13
   |
LL | fn main() { f() }
   |             ^ not found in this scope
   |
note: found an item that was configured out
  --> $DIR/nested-cfg-attrs.rs:2:4
   |
LL | fn f() {}
   |    ^
note: the item is gated here
  --> $DIR/nested-cfg-attrs.rs:1:35
   |
LL | #[cfg_attr(all(), cfg_attr(all(), cfg(FALSE)))]
   |                                   ^^^^^^^^^^
```

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r​? <reviewer name>
-->
<!-- homu-ignore:end -->

estebank committed:
Detect more `cfg`d out items in resolution errors

Use a visitor to collect *all* items (including those nested) that were stripped behind a `cfg` condition.

```
error[E0425]: cannot find function `f` in this scope
  --> $DIR/nested-cfg-attrs.rs:4:13
   |
LL | fn main() { f() }
   |             ^ not found in this scope
   |
note: found an item that was configured out
  --> $DIR/nested-cfg-attrs.rs:2:4
   |
LL | fn f() {}
   |    ^
note: the item is gated here
  --> $DIR/nested-cfg-attrs.rs:1:35
   |
LL | #[cfg_attr(all(), cfg_attr(all(), cfg(FALSE)))]
   |                                   ^^^^^^^^^^
```

Urgau commented on compiler/rustc_expand/src/expand.rs:
```diff
@@ -1260,25 +1260,33 @@ impl InvocationCollectorNode for P<ast::Item> {
-                    ast::UseTreeKind::Nested { items, .. } => {
-                        for (ut, _) in items {
-                            collect_use_tree_leaves(ut, idents);
+        struct ItemNameVisitor(Vec<Ident>);
```
What happens in the case of a double or more stripped items? Won't this visitor and method record them twice?

```rust
#![allow(unexpected_cfgs)]

#[cfg(foo)]
mod fs {
    #[cfg(bar)]
    mod unix {
        fn fchown() {}
    )
}


fn main() {
    fs::unix::chown();
}
```

estebank replied:
This gives you

```
error[E0433]: failed to resolve: use of undeclared crate or module `fs`
  --> f72.rs:13:5
   |
13 |     fs::unix::chown();
   |     ^^ use of undeclared crate or module `fs`
   |
note: found an item that was configured out
  --> f72.rs:4:5
   |
4  | mod fs {
   |     ^^
note: the item is gated here
  --> f72.rs:3:1
   |
3  | #[cfg(foo)]
   | ^^^^^^^^^^^
```

because `fs` is not found so that's the only one that would be checked for (not `unix` or `chown`).

If you check for `unix`, it *will* point at `cfg(foo)`, which is not enough to get `unix` back, but gets you on the way to it:

```
error[E0433]: failed to resolve: use of undeclared crate or module `unix`
  --> f72.rs:14:5
   |
14 |     unix::chown();
   |     ^^^^ use of undeclared crate or module `unix`
   |
note: found an item that was configured out
  --> f72.rs:6:9
   |
6  |     mod unix {
   |         ^^^^
note: the item is gated here
  --> f72.rs:3:1
   |
3  | #[cfg(foo)]
   | ^^^^^^^^^^^
```

In testing, I noticed that we don't point at `fs::unix` when trying `unix` directly and it is behind a `cfg` (the second error):

```
error[E0433]: failed to resolve: could not find `unix` in `fs`
  --> f72.rs:13:9
   |
13 |     fs::unix::chown();
   |         ^^^^ could not find `unix` in `fs`
   |
note: found an item that was configured out
  --> f72.rs:6:9
   |
6  |     mod unix {
   |         ^^^^
note: the item is gated here
  --> f72.rs:5:5
   |
5  |     #[cfg(bar)]
   |     ^^^^^^^^^^^

error[E0433]: failed to resolve: use of undeclared crate or module `unix`
  --> f72.rs:14:5
   |
14 |     unix::chown();
   |     ^^^^ use of undeclared crate or module `unix`
```

If it isn't `cfg`d out *and* it's public, we do provide this suggestion:

```
error[E0433]: failed to resolve: use of undeclared crate or module `unix`
  --> f72.rs:14:5
   |
14 |     unix::chown();
   |     ^^^^ use of undeclared crate or module `unix`
   |
help: consider importing this module
   |
4  + use fs::unix;
   |
```

but if it is private and crate local maybe we should at least point at it (we don't):

```
error[E0433]: failed to resolve: use of undeclared crate or module `unix`
  --> f72.rs:14:5
   |
14 |     unix::chown();
   |     ^^^^ use of undeclared crate or module `unix`
```

estebank replied:
Changed to behave as you expected.

Noratrieb commented on tests/ui/cfg/diagnostics-reexport.rs:
```diff
@@ -1,7 +1,7 @@
 pub mod inner {
-    #[cfg(FALSE)]
+    #[cfg(FALSE)] //~ NOTE the item is gated here
```
I don't think this is correct? no one ever references this item.

estebank replied:
This is because `uwu` under `inner::gone` now gets included for `inner` (because `gone` is `cfg`d out) when trying to access `inner::uwu`. I believe this is the right thing to do. If the user had written `inner::gone::uwu`, the output is:

```
error[E0433]: failed to resolve: could not find `gone` in `inner`
  --> tests/ui/cfg/diagnostics-reexport.rs:38:12
   |
38 |     inner::gone::uwu(); //~ ERROR cannot find function
   |            ^^^^ could not find `gone` in `inner`
   |
note: found an item that was configured out
  --> tests/ui/cfg/diagnostics-reexport.rs:3:9
   |
3  |     mod gone {
   |         ^^^^
note: the item is gated here
  --> tests/ui/cfg/diagnostics-reexport.rs:2:5
   |
2  |     #[cfg(FALSE)] //~ NOTE the item is gated here
   |     ^^^^^^^^^^^^^
```

but if it is `inner::gon::uwu` then the output is

```
error[E0433]: failed to resolve: could not find `gon` in `inner`
  --> tests/ui/cfg/diagnostics-reexport.rs:38:12
   |
38 |     inner::gon::uwu(); //~ ERROR cannot find function
   |            ^^^ could not find `gon` in `inner`
```

We might actually want to still search for `cfg`d out items both with levenshtein distance (but the signal there is much weaker) and by looking at the very last path segment (to find the `cfg`d out `uwu`).

cjgillot replied:
I tend to agree with @Noratrieb: the user wrote `inner::uwu`, which refers to the `use super::uwu` below, which already failed to resolve without suggesting `gone`. Either both suggest `gone::uwu`, or none of them.

Noratrieb commented on tests/ui/cfg/diagnostics-same-crate.rs:
```diff
@@ -37,8 +37,8 @@ mod placeholder {
-#[cfg(i_dont_exist_and_you_can_do_nothing_about_it)]
-pub fn vanished() {}
+#[cfg(i_dont_exist_and_you_can_do_nothing_about_it)] //~ NOTE the item is gated here
+pub fn vanished() {} //~ NOTE found an item that was configured out
```
please update the comment below

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 95c2b4611c9d6a04df0b88a86359c6a07d067728 with merge 623e6c7b5c223cff0a4b75919463e800e0563a50...
<!-- homu: {"type":"TryBuildStarted","head_sha":"95c2b4611c9d6a04df0b88a86359c6a07d067728","merge_sha":"623e6c7b5c223cff0a4b75919463e800e0563a50"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10475531540/job/29017734090)
Build commit: 623e6c7b5c223cff0a4b75919463e800e0563a50 (`623e6c7b5c223cff0a4b75919463e800e0563a50`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10475531540/job/29017734090"},"merge_sha":"623e6c7b5c223cff0a4b75919463e800e0563a50"} -->

rust-timer commented:
Queued 623e6c7b5c223cff0a4b75919463e800e0563a50 with parent a971212545766fdfe0dd68e5d968133f79944a19, future [comparison URL](https://perf.rust-lang.org/compare.html?start=a971212545766fdfe0dd68e5d968133f79944a19&end=623e6c7b5c223cff0a4b75919463e800e0563a50).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (623e6c7b5c223cff0a4b75919463e800e0563a50): [comparison URL](https://perf.rust-lang.org/compare.html?start=a971212545766fdfe0dd68e5d968133f79944a19&end=623e6c7b5c223cff0a4b75919463e800e0563a50&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=a971212545766fdfe0dd68e5d968133f79944a19&end=623e6c7b5c223cff0a4b75919463e800e0563a50&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2%  | [0.2%, 0.2%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 1.9%  | [1.9%, 1.9%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.2% | [-0.3%, -0.2%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.1% | [-0.3%, 0.2%]  | 3     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=a971212545766fdfe0dd68e5d968133f79944a19&end=623e6c7b5c223cff0a4b75919463e800e0563a50&stat=max-rss)
<details>
<summary>Results (primary 2.6%, secondary 1.5%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.6%  | [1.8%, 3.3%]   | 3     |
| Regressions ❌ <br /> (secondary)  | 3.1%  | [1.7%, 4.3%]   | 4     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -4.8% | [-4.8%, -4.8%] | 1     |
| All ❌✅ (primary)                 | 2.6%  | [1.8%, 3.3%]   | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=a971212545766fdfe0dd68e5d968133f79944a19&end=623e6c7b5c223cff0a4b75919463e800e0563a50&stat=cycles:u)
<details>
<summary>Results (secondary -2.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.9% | [-3.1%, -2.6%] | 6     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=a971212545766fdfe0dd68e5d968133f79944a19&end=623e6c7b5c223cff0a4b75919463e800e0563a50&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.1% | [0.0%, 0.4%] | 83    |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.1% | [0.0%, 0.4%] | 83    |
</details>

**Bootstrap**: 749.153s -> 748.829s (-0.04%)
**Artifact size**: 338.89 MiB -> 338.83 MiB (-0.02%)

estebank commented:
It might make sense to limit the visitor to two levels of nesting at most, as a way to attempt to reduce the impact on crates that have large modules `cfg`d out. I suspect that we might *not* have any test case of a large crate in that configuration, and that's why we're not seeing more effects from this change.

estebank committed:
Limit how deep we visit items to find cfg'd out names

estebank commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 0e22c731042bf91dcb713f2cefcb6477b281441d with merge 811a69609bf7d38bbab02fa8f4d3b4cb98802f24...
<!-- homu: {"type":"TryBuildStarted","head_sha":"0e22c731042bf91dcb713f2cefcb6477b281441d","merge_sha":"811a69609bf7d38bbab02fa8f4d3b4cb98802f24"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10606006671/job/29398970885)
Build commit: 811a69609bf7d38bbab02fa8f4d3b4cb98802f24 (`811a69609bf7d38bbab02fa8f4d3b4cb98802f24`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10606006671/job/29398970885"},"merge_sha":"811a69609bf7d38bbab02fa8f4d3b4cb98802f24"} -->

rust-timer commented:
Queued 811a69609bf7d38bbab02fa8f4d3b4cb98802f24 with parent 100fde5246bf56f22fb5cc85374dd841296fce0e, future [comparison URL](https://perf.rust-lang.org/compare.html?start=100fde5246bf56f22fb5cc85374dd841296fce0e&end=811a69609bf7d38bbab02fa8f4d3b4cb98802f24).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.8 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (811a69609bf7d38bbab02fa8f4d3b4cb98802f24): [comparison URL](https://perf.rust-lang.org/compare.html?start=100fde5246bf56f22fb5cc85374dd841296fce0e&end=811a69609bf7d38bbab02fa8f4d3b4cb98802f24&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=100fde5246bf56f22fb5cc85374dd841296fce0e&end=811a69609bf7d38bbab02fa8f4d3b4cb98802f24&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2%  | [0.2%, 0.3%]   | 4     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.3% | [-0.3%, -0.3%] | 1     |
| All ❌✅ (primary)                 | 0.2%  | [0.2%, 0.3%]   | 4     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=100fde5246bf56f22fb5cc85374dd841296fce0e&end=811a69609bf7d38bbab02fa8f4d3b4cb98802f24&stat=max-rss)
<details>
<summary>Results (primary 1.6%, secondary 2.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.6% | [0.6%, 2.6%] | 4     |
| Regressions ❌ <br /> (secondary)  | 2.3% | [2.1%, 2.7%] | 3     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.6% | [0.6%, 2.6%] | 4     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=100fde5246bf56f22fb5cc85374dd841296fce0e&end=811a69609bf7d38bbab02fa8f4d3b4cb98802f24&stat=cycles:u)
<details>
<summary>Results (secondary -2.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.9% | [-3.4%, -2.6%] | 14    |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=100fde5246bf56f22fb5cc85374dd841296fce0e&end=811a69609bf7d38bbab02fa8f4d3b4cb98802f24&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.1% | [0.0%, 0.5%] | 83    |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.1% | [0.0%, 0.5%] | 83    |
</details>

**Bootstrap**: 753.9s -> 754.227s (0.04%)
**Artifact size**: 338.87 MiB -> 339.03 MiB (0.05%)

estebank commented:
I suspect that the regressions on incr-comp for html5ever-0.26 are caused by [this gated macro expansion](https://docs.rs/crate/html5ever/0.26.0/source/src/tokenizer/mod.rs#604-610). [clap-3.1.6 is quite feature heavy](https://docs.rs/crate/clap/3.1.6/source/Cargo.toml#317-333). I leave it to you whether the impact on incr-comp of munching through more nodes is worth it. I believe it it is, as when items are `cfg`d out it leads to very confusing output otherwise.

joshka commented:
Possibly related enough to consider as part of this (if it's not already). https://github.com/rust-lang/rust/issues/134448
(TLDR - cfg out derive macro and trait with the same path don't trigger this diagnostic)

Noratrieb commented on tests/ui/cfg/diagnostics-same-crate.rs:
```diff
@@ -49,18 +49,16 @@ fn main() {
-    // but currently don't have one.
-    // Not that it matters much though, this is highly unlikely to confuse anyone.
+    // Exists in the crate root - we show a diagnostic because we treat "no module DefId" as "crate
+    // root DefId".
```
I don't think this is correct? it should be the current module def id instead

estebank replied:
It's been a while since I last looked at this code, but *I think* this was referring to https://github.com/rust-lang/rust/pull/129183/files#diff-0e459b4cbb41dd93274d7c68b18a2a3eab7152e9aea107839a7bddff0b4e9195R819

Noratrieb replied:
Oh yeah I didn't clarify, the comment matches the implementation but I think the implementation is wrong

Noratrieb commented:
@estebank can you explain why this nesting is necessary? I fail to see a case where nesting is necessary, as resolution will always fail at the outermost missing item, making the inner ones physically inaccessible.

estebank committed:
remove recursive search for items

estebank commented:
> I fail to see a case where nesting is necessary, as resolution will always fail at the outermost missing item, making the inner ones physically inaccessible.

Removed the recursive search. My logic was "the user already thinks an item exists, they might have both used the wrong path *and* have it disabled behind a cfg flag" where knowing the item is somewhere but not seeing the cfg would be confusing. But left with a single level as I think we should land at least the simpler version.

bors commented:
:umbrella: The latest upstream changes (presumably #133154) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #138740) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

apiraino commented:
@Noratrieb @Urgau given you left some comments, do you have capacity to help reviewing here? I am going to reroll the review assignment but feel free to take it :slightly_smiling_face: thanks

r? compiler

estebank committed:
tiny cleanup

estebank committed:
Tweak rendering of cfg'd out item

```
error[E0433]: failed to resolve: could not find `doesnt_exist` in `inner`
  --> $DIR/diagnostics-cross-crate.rs:18:23
   |
LL |     cfged_out::inner::doesnt_exist::hello();
   |                       ^^^^^^^^^^^^ could not find `doesnt_exist` in `inner`
   |
note: found an item that was configured out
  --> $DIR/auxiliary/cfged_out.rs:6:13
   |
LL |     #[cfg(false)]
   |           ----- the item is gated here
LL |     pub mod doesnt_exist {
   |             ^^^^^^^^^^^^
```

estebank commented:
@bors r=davidtwco

bors commented:
:pushpin: Commit 4b24c4bf23df8ae5c53669e3209b9f3074769b69 has been approved by `davidtwco`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=davidtwco 4b24c4bf23df8ae5c53669e3209b9f3074769b69 -->
<!-- homu: {"type":"Approved","sha":"4b24c4bf23df8ae5c53669e3209b9f3074769b69","approver":"davidtwco","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 4b24c4bf23df8ae5c53669e3209b9f3074769b69 with merge 889701db1ff614160314734fe4138c2f820a95bb...
<!-- homu: {"type":"BuildStarted","head_sha":"4b24c4bf23df8ae5c53669e3209b9f3074769b69","merge_sha":"889701db1ff614160314734fe4138c2f820a95bb"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/rust/actions/runs/16690186899/job/47250138447)
Approved by: davidtwco
Pushing 889701db1ff614160314734fe4138c2f820a95bb to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"davidtwco","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/rust/actions/runs/16690186899/job/47250138447"},"merge_sha":"889701db1ff614160314734fe4138c2f820a95bb"} -->
